src/js/common/validate.js

```diff
@@ -0,0 +1,343 @@
+
+  This function handles the general case as well as escaped variables, and converts them to a list of path segments
+  like `["sky", "middle.earth", "moria"]`.  These segments are used in other functions to check for the existence of
+  variables, etc.
+
+ */
+gpii.schema.validator.extractPathSegments = function (string) {
+    var segments = [];
+
+    // A regular expression to split the current segment and the remainder from one another. Handles variations like:
+    //
+    // 1. `simple.simon`
+    // 2. `['kinda.complex'].simple`
+    // 3. `simple['kinda.complex']`
+    // 4. `['really.complex']['don\'t I know it']`
+    //
+    // See https://xkcd.com/1171/
+    //
+    // The matching within single quotes must be non-greedy to ensure that things like `['a.b']['c.d']` are treated as
+    // separate segments.
+    var slicingRegexp = /^\.?(\['.+?']|[^\[\.]+)([\[\.].+)$/;
+
+    var remainingPath = string;
+
+    // Iterate through, splitting by dots while preserving escaped dot notation (see above).
+    var matches = remainingPath.match(slicingRegexp);
+    while (matches) {
+        segments.push(gpii.schema.validator.sanitizePathSegment(matches[1]));
+        remainingPath = matches[2];
+        matches = remainingPath.match(slicingRegexp);
+    }
+
+    // The last segment will not have two parts and can be added in its entirety
+    segments.push(gpii.schema.validator.sanitizePathSegment(remainingPath));
+
+    return segments;
+};
+
+/*
+
+  Convenience function to extract the path segments from the `error` data structure returned by AJV.
+
+ */
+gpii.schema.validator.extractPathSegmentsFromError = function (error) {
+    return gpii.schema.validator.extractPathSegments(error.dataPath);
+};
+
+
+/*
+
+  Because dots are allowed in Javascript/JSON object keys, any segment with a dot is escaped by AJV.  The simplest
+  example is `this.that`, which would be escaped as `['this.that']`.  This gives us three additional special characters,
+  namely the square brackets and the single quote.
+
+  If any of those are used, additional escaping is introduced, so `['this.that']` becomes `['[\\'this.that\\']']`.  This
+  introduces the final special character, namely the backslash.
+
+  If an escaped backslash is used, it is double escaped, so that `a\\b` becomes `a\\\\b`.
+
+  This function takes care of reversing this process including all combinations, to ensure that we end up with the
+  original unescaped literal key.
+
+ */
+gpii.schema.validator.sanitizePathSegment = function (segment) {
+    // Discard any leading dot
+    var segmentMinusLeadingDot = segment.replace(/^\./, "");
+
+    // If we are surrounded by `['']`, extract the inner content and then unescape it.
+    var hasSpecialRegexp = /\['(.+)'\]/;
+    var specialMatches = segmentMinusLeadingDot.match(hasSpecialRegexp);
+    if (specialMatches) {
+        // unescape the remaining variables by using them to create a new `String`.
+        return String(specialMatches[1]);
```

Comment:
This may have been a bit of voodoo left over from a previous implementation, all tests pass if we simply return the inner matched value, so I have made that change.